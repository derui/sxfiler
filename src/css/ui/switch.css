.switch {
  position: relative;
  display: inline-block;
  background-color: transparent;
  width: 32px;
  max-width: 32px;
  height: 20px;
}

.switch__rail {
  position: relative;
  display: inline-block;
  margin: 0;
  padding: 0 4px;

  /* background-color: rgb(var(--color-base03)); */
  background-color: var(--color-ui-switchRail);
  width: 100%;
  height: 100%;
  transition: background-color ease-out 100ms;

  &.switch__rail--checked {
    /* background-color: rgb(var(--color-blue)); */
    background-color: var(--color-ui-switchRailChecked);
  }
}

.switch__box {
  position: absolute;
  display: inline-block;

  /* background-color: rgb(var(--color-base3)); */
  background-color: var(--color-ui-switchBox);
  height: 100%;
  width: 16px;
  left: 0;
  top: 0;
  box-shadow: -2px 2px 2px 2px rgba(0, 0, 0, 0.2);
  transition: transform ease-out 100ms;

  &.switch__box--checked {
    transform: translateX(calc(32px - 16px + 8px));
  }
}

.switch__input {
  display: none;
}
