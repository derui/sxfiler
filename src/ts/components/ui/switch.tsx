import { h } from "preact";
import { useState, StateUpdater, useRef } from "preact/hooks";

type Callback = (value: boolean) => void;
export type Props = {
  onChange: Callback;
  checked?: boolean;
};

const handleChange = (updateState: StateUpdater<boolean>, callback: Callback) => (e: Event) => {
  const target = e.target as HTMLInputElement;
  const v = target.checked;

  updateState(v);
  callback(v);
};

export const Component: preact.FunctionalComponent<Props> = (props) => {
  const { onChange, checked = false } = props;
  const [state, updateState] = useState(checked);
  // for clicking checkbox directly
  const ref = useRef<HTMLInputElement>();

  const boxClass = `
switch__box
${state ? "switch__box--checked" : ""}
`;

  const railClass = `
switch__rail
${state ? "switch__rail--checked" : ""}
`;

  return (
    <span
      class="switch"
      data-testid="switch-root"
      onClick={() => {
        if (ref.current) {
          ref.current.click();
        }
      }}
    >
      <span class={railClass}>
        <span class={boxClass}></span>
      </span>
      <input
        ref={ref}
        class="switch__input"
        type="checkbox"
        checked={state}
        onChange={handleChange(updateState, onChange)}
      />
    </span>
  );
};
