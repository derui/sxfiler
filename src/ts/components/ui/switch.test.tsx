import { h } from "preact";
import { render as renderAsText } from "preact-render-to-string";
import { render, fireEvent, act } from "@testing-library/preact";
import { Component } from "./switch";

describe("UI", () => {
  describe("Switch", () => {
    test("display unchecked when property is not given", () => {
      const tree = renderAsText(<Component onChange={() => {}} />);

      expect(tree).toMatchSnapshot();
    });

    test("display unchecked when property is not given", () => {
      const tree = renderAsText(<Component checked={true} onChange={() => {}} />);

      expect(tree).toMatchSnapshot();
    });

    test("get changed value when click it", async () => {
      const callback = jest.fn();
      const query = render(<Component checked={true} onChange={callback} />);
      const root = await query.findByTestId("switch-root");

      act(() => {
        fireEvent.click(root);
      });

      expect(callback).toBeCalledWith(false);
    });

    test("get changed value if default value is not given", async () => {
      const callback = jest.fn();
      const query = render(<Component onChange={callback} />);
      const root = await query.findByTestId("switch-root");

      act(() => {
        fireEvent.click(root);
      });

      expect(callback).toBeCalledWith(true);
    });
  });
});
