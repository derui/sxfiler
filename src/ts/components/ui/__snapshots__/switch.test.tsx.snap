// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Switch display unchecked when property is not given 1`] = `
"<span class=\\"switch\\" data-testid=\\"switch-root\\"><span class=\\"
switch__rail

\\"><span class=\\"
switch__box

\\"></span></span><input class=\\"switch__input\\" type=\\"checkbox\\" /></span>"
`;

exports[`UI Switch display unchecked when property is not given 2`] = `
"<span class=\\"switch\\" data-testid=\\"switch-root\\"><span class=\\"
switch__rail
switch__rail--checked
\\"><span class=\\"
switch__box
switch__box--checked
\\"></span></span><input class=\\"switch__input\\" type=\\"checkbox\\" checked /></span>"
`;
